### PR TITLE
Make authority discovery configurable

### PR DESCRIPTION
**Pull request by bkchr**
This pr makes the authority discovery configurable. So, instead of
having default values for the query interval, publish interval etc this
pr adds a configuration to make these values changeable. This will be
useful for tests where authority discovery is required.

Closes: https://github.com/paritytech/substrate/issues/7462

### CODE REVIEW

**Review by mxinden — COMMENTED** [review]
Do users need to fine tune each value, or is there really only the need for either running in `normal` or `testing` mode? If the latter is the case, I would prefer not exposing all these internals to the outside, but instead only offer to configure one of the two modes, which under the hood set the corresponding configuration settings.

**Comment by bkchr** [comment]
All these values are currently fine tuned for Kusama(based on the comments in the code). Why not make all these values configurable? I don't see the problem? 

I don't want to get back here to change every value because I realize later that the test mode isn't configured correctly.

File: client/authority-discovery/src/worker.rs [hunk 1]
```diff
@@ -154,30 +151,31 @@ where
 		dht_event_rx: DhtEventStream,
 		role: Role,
 		prometheus_registry: Option<prometheus_endpoint::Registry>,
+		config: crate::WorkerConfig,
 	) -> Self {
-		// Kademlia's default time-to-live for Dht records is 36h, republishing records every 24h.
+		// Kademlia's default time-to-live for Dht records is 36h, republishing records every 12h by default.
```
Comment (mxinden) [inline client/authority-discovery/src/worker.rs:156]:
```suggestion
		// Kademlia's default time-to-live for Dht records is 36h, republishing records every 24h through libp2p-kad.
```

`libp2p-kad` republishes values every 24h. In addition, the authority discovery module re-triggers publishing every 12h to make sure values persist across restarts.

See [libp2p-kad/behaviour](https://github.com/libp2p/rust-libp2p/blob/421ae9d9c9cfe1398f8bf6e2dd7d90c8e8f36018/protocols/kad/src/behaviour.rs#L257-L269) for details.

File: client/authority-discovery/src/lib.rs [hunk 2]
```diff
@@ -43,6 +43,48 @@ mod service;
...
+pub struct WorkerConfig {
+	/// The interval in which the node will publish its own address on the DHT.
+	///
+	/// By default this is set to 12 hours.
+	pub publish_interval: Duration,
+	/// The interval in which the node will query the DHT for new entries.
+	///
+	/// By default this is set to 10 minutes.
+	pub query_interval: Duration,
+	/// The time the node will wait before doing the first query of the DTH interval.
```
Comment (mxinden) [inline client/authority-discovery/src/lib.rs:56]:
```suggestion
	/// The time the node will wait before triggering the first DHT query or publish.
```

File: client/authority-discovery/src/lib.rs [hunk 3]
```diff
@@ -43,6 +43,48 @@ mod service;
...
+	///
+	/// By default this is set to 10 minutes.
+	pub priority_group_set_interval: Duration,
+	/// The time the worker will wait before instructing the peerset the first time.
+	///
+	/// Be aware that the actual delay will be computed by [`Self::query_start_delay`] +
+	/// [`Self::priority_group_set_start_delay`]
+	///
+	/// By default this is set to 5 minutes.
+	pub priority_group_set_start_delay: Duration,
```
Comment (mxinden) [inline client/authority-discovery/src/lib.rs:73]:
```suggestion
	pub priority_group_set_offset: Duration,
```

File: client/authority-discovery/src/lib.rs [hunk 4]
```diff
@@ -43,6 +43,48 @@ mod service;
...
+	/// By default this is set to 30 seconds.
+	///
+	/// This default is based on the rough boostrap time required by libp2p Kademlia.
+	pub query_start_delay: Duration,
+	/// The interval in which the worker will instruct the peerset to connect to a random subset
+	/// of discovered validators.
+	///
+	/// By default this is set to 10 minutes.
+	pub priority_group_set_interval: Duration,
+	/// The time the worker will wait before instructing the peerset the first time.
```
Comment (mxinden) [inline client/authority-discovery/src/lib.rs:67]:
```suggestion
	/// The time the worker will wait after each query interval tick to pass a subset of the cached authority addresses down to the peerset.
```

**Review by mxinden — COMMENTED** [review]
> All these values are currently fine tuned for Kusama(based on the comments in the code). Why not make all these values configurable? I don't see the problem?

Exposing these configuration options to the outside of the module makes them part of its public contract. In my eyes there are two consequences:

1. A change to either of these options default values is a breaking change.

2. They are internal concepts which require knowledge of the inner working of the module. External users configuring the module can easily make mistakes.

> I don't want to get back here to change every value because I realize later that the test mode isn't configured correctly.

Do I understand correctly that this change is needed for unit / integration tests only? If so, could you not overwrite your Substrate path in Polkadot, fine tune the values for your tests to succeed, and then include the values in this pull request?

**Comment by bkchr** [comment]
> > All these values are currently fine tuned for Kusama(based on the comments in the code). Why not make all these values configurable? I don't see the problem?
> 
> Exposing these configuration options to the outside of the module makes them part of its public contract. In my eyes there are two consequences:
> 
>     1. A change to either of these options default values is a breaking change.
> 
>     2. They are internal concepts which require knowledge of the inner working of the module. External users configuring the module can easily make mistakes.
> 

I understand this and I can emphasize this more in the docs that people should only change these values when they are sure what they doing. Substrate is not only Polkadot/Kusama and will it is okay to have default values tuned for them, what should other chains do which may require a different setup? 

Configuring stuff always requires that you have a deeper understanding on what the values mean and do, but if you want you can do it. Why not give people freedom?
image

> 
> > I don't want to get back here to change every value because I realize later that the test mode isn't configured correctly.
> 
> Do I understand correctly that this change is needed for unit / integration tests only? If so, could you not overwrite your Substrate path in Polkadot, fine tune the values for your tests to succeed, and then include the values in this pull request?

See what I have written above, but also there will be integration tests and later tests using "real" nodes for which we may also want to have different values.

**Review by mxinden — APPROVED** [review]
The changes in itself look good to me. Thank you for the detailed doc comments.

I am still reluctant to exposing the full range of configuration options to the user. Instead of exposing all of them because a user _might_ need them in the future, I would prefer to delay this decision to a point in time where we know more on the concrete use-case. As far as I can tell for today's use-case, using authority discovery in integration tests, the `Mode` suggestion above would suffice.

That said, I don't think my opinion should be blocking this pull request.

**Pull request merged**